### PR TITLE
[Debug] Create a main guide for Debugging

### PR DESCRIPTION
**Pull request by wouterj**
This topic was missing the main guide. This PR fixes that and also adds a couple more debugging tips.

/fixes https://github.com/symfony/symfony-docs/issues/6342

### CODE REVIEW

File: debug.rst [hunk 1]
```diff
@@ -1,8 +1,136 @@
-Debugging
-=========
+.. index::
+   single: Debugging
 
-.. toctree::
-    :maxdepth: 1
-    :glob:
+How to Optimize your Development Environment for Debugging
```
Comment (xabbuh) [inline debug.rst]: "Your"?

File: debug.rst [hunk 2]
```diff
@@ -1,8 +1,136 @@
...
+==========================================================
 
-    debug/*
+When you work on a Symfony project on your local machine, you should use the
+``dev`` environment (``app_dev.php`` front controller). This environment
+configuration is optimized for two main purposes.
```
Comment (xabbuh) [inline debug.rst]: End this paragraph with a colon?

File: debug.rst [hunk 3]
```diff
@@ -1,8 +1,136 @@
...
+
+* Give the developer accurate feedback whenever something goes wrong (provided
+  by the web debug toolbar, nice exception pages, profiler, ...);
+* Be as similar as possible as the production environment to avoid problems
```
Comment (xabbuh) [inline debug.rst]: "Be as close as possible to the production [...]"?

File: debug.rst [hunk 4]
```diff
@@ -1,8 +1,136 @@
...
+  when deploying the project.
+
+Using Interactive Debug Tools
+-----------------------------
+
+Interactive debug tools allow you to walk through the code step by step,
+making it easier to indentify which step is causing problems. Symfony works
```
Comment (xabbuh) [inline debug.rst]: identify

File: debug.rst [hunk 5]
```diff
@@ -1,8 +1,136 @@
...
+with any PHP debug environment, among them:
+
+ * Xdebug_, the most well-known PHP debugger;
+ * PsySH_, a PHP REPL_ (Read-eval-print loop) debugger. Use the
```
Comment (xabbuh) [inline debug.rst]: Don't we always enclose links with backticks?

File: debug.rst [hunk 6]
```diff
@@ -1,8 +1,136 @@
...
+    $loader = require_once __DIR__.'/../app/autoload.php';
+    require_once __DIR__.'/../app/AppKernel.php';
+
+    $kernel = new AppKernel('dev', true);
+    // $kernel->loadClassCache();
+    $request = Request::createFromGlobals();
+
+.. tip::
+
+    If you disable the PHP caches, don't forget to revert after your debugging
```
Comment (xabbuh) [inline debug.rst]: [...] revert these changes after [...]

File: debug.rst [hunk 7]
```diff
@@ -1,8 +1,136 @@
...
+
+    $kernel->loadClassCache('classes', '.php.cache');
+
+Dumping Variables with the VarDumper
+------------------------------------
+
+To ease the debugging of a variable in your application, you can use the
+:doc:`VarDumper component </components/var_dumper>` to dump the content of a
+variable. The component provides an alternative to the PHP :phpfunction:`var_dump()`
+function, in the form of ``dump()``::
```
Comment (xabbuh) [inline debug.rst]: [...] provides the `dump()` function, an alternative to PHP's `var_dump()` function:

File: debug.rst [hunk 8]
```diff
@@ -1,8 +1,136 @@
...
+    // create a variable with a value
+    $myVar = ...;
+
+    // and dump it
+    dump($myVar);
+
+The dumper is not limited to scalar values. Arrays and objects can also be
+visualized using the VarDumper. One of the most important advantages of using
+``dump()`` is a nicer and more specialized dump of objects (e.g. Doctrine
+internals are filtered out when dumping a proxy entity).
```
Comment (xabbuh) [inline debug.rst]: [...] dumping an entity proxy

File: debug.rst [hunk 9]
```diff
@@ -1,8 +1,136 @@
...
+
+    {# dumps the variable to the web debug toolbar to not modify the template #}
+    {% dump myVar %}
+
+Useful Debugging Commands
+-------------------------
+
+When developing a large application, it can be hard to keep track of all the
+different services, routes and translations. Luckily, Symfony has some commands
+that can help you visualize and find the information.
```
Comment (xabbuh) [inline debug.rst]: End the paragraph with a colon?

File: debug.rst [hunk 10]
```diff
@@ -1,8 +1,136 @@
...
+
+``debug:container``
+    Displays information about the contents of the Symfony container for all public
+    services. To find only those matching a name, append the name as an argument.
+
+``debug:config``
+    Shows all configured bundles, their class and their alias.
```
Comment (xabbuh) [inline debug.rst]: [...] their classes and their aliases.

File: debug.rst [hunk 11]
```diff
@@ -1,8 +1,136 @@
...
+
+``debug:translation <locale>``
+    Shows a table of the translation key, the domain, the translation and the
+    fallback translation for all known messages, if translations exist for
+    the given locale.
+
+.. tip::
+
+    When in doubt how to use a console command, open the help section by
+    appending the ``--help`` option.
```
Comment (xabbuh) [inline debug.rst]: (`-h`)

File: debug.rst [hunk 12]
```diff
@@ -1,8 +1,136 @@
...
+``debug:config``
+    Shows all configured bundles, their class and their alias.
+
+``debug:router``
+    Displays information about all configured routes in the application as a
+    table with the name, method, scheme, host and path for each route.
+
+``debug:translation <locale>``
+    Shows a table of the translation key, the domain, the translation and the
+    fallback translation for all known messages, if translations exist for
```
Comment (xabbuh) [inline debug.rst]: Remove the comma?

**Comment by xabbuh** [comment]
I just left some minor comments, but overall I absolutely like this PR. Great work as always Wouter. 👍 

Status: Reviewed

**Comment by xabbuh** [comment]
Oh, and we need to update the `/configuration/front_controllers_and_kernel.rst` and fix the link (see the failing build job).

File: debug.rst [hunk 13]
```diff
@@ -1,8 +1,136 @@
...
+that can help you visualize and find the information.
+
+``debug:container``
+    Displays information about the contents of the Symfony container for all public
+    services. To find only those matching a name, append the name as an argument.
+
+``debug:config``
+    Shows all configured bundles, their class and their alias.
+
+``debug:router``
```
Comment (javiereguiluz) [inline debug.rst]: Maybe we should reference here the `router:match` command too. In my experience it's very useful for newcomers and is very useful to quickly spot routing problems.

**Comment by wouterj** [comment]
Ready to merge, thanks for the great review and sorry for the loooooong delay!

File: debug.rst [hunk 14]
```diff
@@ -1,8 +1,140 @@
...
+
+    $kernel->loadClassCache('classes', '.php.cache');
+
+Dumping Variables with the VarDumper
+------------------------------------
+
+To ease the debugging of a variable in your application, you can use the
+:doc:`VarDumper component </components/var_dumper>` to dump the content of a
+variable. The component provides the ``dump()`` function, an alternative to
+PHP's :phpfunction:`var_dump()` function::
```
Comment (theofidry) [inline debug.rst]: I would add a not on how to install it or a link for it, as IIRC, unless the application is booted and if you forget to add the `dump.php` file in your `composer.json`, `dump()` won't work

File: debug.rst [hunk 15]
```diff
@@ -1,8 +1,140 @@
...
+.. tip::
+
+    If you disable the PHP caches, don't forget to revert these changes after
+    your debugging session.
+
+Some IDEs do not like the fact that some classes are stored in different
+locations. To avoid problems, you can either tell your IDE to ignore the PHP
+cache files, or you can change the extension used by Symfony for these files::
+
+    $kernel->loadClassCache('classes', '.php.cache');
```
Comment (theofidry) [inline debug.rst]: is `loadClassCache()` still a thing? I though it was no longer useful with PHP7

Comment (wouterj) [reply debug.rst]: Indeed, let's just remove this section.

File: debug.rst (hunk 14 above)
Comment (wouterj) [reply debug.rst]: The section refers to the VarDumper component installation docs. I don't think we have to duplicate the info here.

**Comment by wouterj** [comment]
Ready to merge imo

**Comment by javiereguiluz** [comment]
Wouter, I'm afraid I must close this as "won't fix" because we decided to remove the main debugging article (see #8538) and its contents are moved to different articles of the docs. Cheers!
